### DISCUSSION TITLE
Is there any plan for the future of AIR regarding HUAWEI Harmony OS?

### DISCUSSION THREAD

**Discussion by Jason0247** [airsdk/Adobe-Runtime-Support#2804] [Q&A]
In the near future, HUAWEI Harmony OS may no longer support APK and become an independent system. Do you have any plans for AIR in this regard?
Thank you!

**Comment by marchbold** [comment]
I thought Harmony OS could run any Android apps as long as you don't include any Google services? So AIR should already support it via the aab/apk package output.

**Reply by bilalabbasi09** [reply]
[Huawei](https://exploremobiles.com/brand/huawei/) will no longer supported APK not just a google service.

**Comment by Jason0247** [comment]
Huawei's explanation is that in the future, APK will no longer be supported, not just Google services. @marchbold

**Reply by marchbold** [reply]
Interesting. Have you got a link to this announcement?

**Comment by Jason0247** [comment]
https://baijiahao.baidu.com/s?id=1773592115496538443&wfr=spider&for=pc
@marchbold 
This is the news announced on the Chinese website that in the future, APK will be completely abandoned after version 5.0.

**Reply by marchbold** [reply]
Gosh, bold choice. Going to alienate a lot of existing developers. 

I suppose this is a wait and see situation, not much can be done until they officially announce SDK / development tools etc.

In the meantime, AppGallery and the current HarmonyOS supports android targets.

**Reply by Jason0247** [reply]
It is indeed a very bold thing, a new system, a new ecology, and I personally think it is still very difficult.
But currently, their decision will not be realized in the short term. But if we consider the large number of Chinese users, we don't know if it will really be implemented in the future.
Let's wait and see first.

**Reply by ajwfrost** [reply]
I had a quick look - the OS looks like another one that's very much focused on web applications. Still Linux-based but we'd need to see what they end up with in terms of the application packaging and deployment options.. if they just go with a kind of web runtime option, then it would be a challenge!

In terms of whether we'd support it .. that just depends on the cost/benefit really. If it's a huge effort to do, then it's less likely to happen, unless we can be fairly sure of getting our investment costs back via the AIR SDK license sales. (So like with Linux where we restricted its use to just people who at least are paying something towards it).

Of course, there's always the option for privately funded integrations, rather than releasing it as a generic supported platform..

**Reply by RayGyoe** [reply]
To support this environment, AIR should study ArkCompiler ： https://developer.harmonyos.com/en/docs/documentation/doc-guides/build_overview-0000001055075201

**Reply by ajwfrost** [reply]
Thanks - so that part looks quite similar to how we build the runtimes for Android currently, with gradle being the main tool used and then this kicking off native compilation via cmake.

For applications that then used this, we'd need to be able to import the pre-built native libraries.

It's then just a case of knowing what we need to use in place of all the Android APIs for all the device interactions, video decoding, native window access etc... So still could be a fair amount of effort :-(

thanks

**Reply by RayGyoe** [reply]
Yes, it requires a lot of effort. Investment may not necessarily yield corresponding returns.

I prefer to improve performance without changing any content, rather than supporting a new ecosystem

**Comment by wmtj** [comment]
Huawei HarmonyOS is already running on more than a billion devices (including iot devices). @ajwfrost I think you should learn more about it.  :)

On existing platforms (Android/iOS...) AIR has become a de facto maintenance tool for older projects, rather than an option for newer ones. New platforms, new opportunities. If AIR does not actively support this new platform and vast ecosystem, it is likely that AIR will miss a major opportunity for future growth. 

HarmonyOS will drop support for apk after version 5.0, and then I think I'll have to choose a new development platform if AIR does not support it.

**Reply by liangqi** [reply]
HarmonyOS 5 devices exceed 10 million mark, reveals Huawei official - July 30, 2025
https://www.huaweicentral.com/harmonyos-5-devices-exceed-10-million-mark-reveals-huawei-official/

https://en.wikipedia.org/wiki/HarmonyOS_NEXT
The kernel of HarmonyOS NEXT no longer includes the compatibility layer of [AOSP](https://en.wikipedia.org/wiki/Android_(operating_system)) framework with Android libraries from [EMUI](https://en.wikipedia.org/wiki/EMUI) in the [user space](https://en.wikipedia.org/wiki/User_space_and_kernel_space) and cannot run Android [apk](https://en.wikipedia.org/wiki/Apk_(file_format)) apps natively, as was the case with the initial dual-framework HarmonyOS

"The billion devices(including iot devices)" can't run HarmonyOS 5.0 or next.

**Reply by ajwfrost** [reply]
We would still like to be able to support this. Things are becoming a little clearer but we don't have any way to test things currently - we will need to get hold of one of these devices I guess. The emulator doesn't support OpenGL ES calls (and it's still not 100% clear whether there is appropriate support across the APIs that we normally use, the documentation is still not particularly clear - especially whether it refers to "HarmonyOS NEXT 5" or earlier versions, or OpenHarmony (which I gather is a bit different?).

We did buy a Huawei tablet that ran on HarmonyOS but it turned out to be version 3 - plus we couldn't get it to connect to a MacBook via ADB, in case we were able to try running Harmony-native apps on it...

I've just checked Huawei's UK retail site and their latest tablet comes with HarmonyOS 4.3. Again, not sure whether we could actually get one of these and use it for development of an app that would then work on HarmonyOS NEXT 5....?

**Reply by liangqi** [reply]
HarmonyOS 5.0(or NEXT) and later are only available inside mainland China, as far as I know. Not sure when they will be released worldwidely.

**Comment by Ender22** [comment]
My company runs a b2b app and has some Chinese clients. Supporting Harmony would be really important for us too

**Comment by ajwfrost** [comment]
Okay thanks -> @wmtj I'll have to do that, I had originally thought it was an Android/AOSP clone that they set up after Google blocked Huawei from the official Android certification process. 

We've just started looking into another platform but potentially it could be more efficient to bring both of them onboard at the same time. I'll ask one of the team to do a bit of investigation..

thanks

**Reply by wmtj** [reply]
Thanks for your reply! @ajwfrost

I have to say that although China's market and developer community is very large, supporting Huawei HarmonyOS is not simply to support Chinese developers, but to support global developers.

Any political and economic sanctions have an end, and since the recent release of Huawei Mate 60 phone, the world has understood that chip sanctions cannot stop the advance of China's high-tech sector. According to reports: the United States is trying to sell previously banned advanced process chips now, and Huawei has once again starting to try to sell phones globally (All chips are made in China, the previous sanctions are ineffective. Even new sanctions comes... eventually the conflict will end in some form of 'reconciliation' ...) Just a few years later, the world's major mobile systems will be like this: iOS, Android, HarmonyOS, XXOS... By then, AIR only support iOS & Android... I do not believe that any company's technology decision maker would consider a technology path that cannot support all major platforms as a rational choice, or even rule it out in the first place. Therefore, my previous reply said that supporting Huawei HarmonyOS is an opportunity for AIR,  "Rise of AIR". Doing some proper commercial design, charging License fees, etc., is acceptable, there are a large number of Flash developers in China, as well as worldwide.

Frankly, I'm very surprised that the AIR team lacks the most basic understanding of HarmonyOS, which is already running on more than a billion devices, and I personally recommend that you at least buy a Huawei phone running HarmonyOS and have someone track and research it. Of course, it's a little late for me to say that.

I'm sorry, my English is poor, some of the above content was translated by machine, I believe you can understand. :)

**Reply by ajwfrost** [reply]
Thanks for your insights - and yes whilst I wasn't much aware about HarmonyOS (other than hearing about the original announcement of this a while ago) I do know that the Flash situation in China is very different, we did a couple of custom builds of the Flash Player a number of years ago and now of course it's still supported by a local gaming company. We'll look at getting a handset, as you suggest, and see what's involved in adding the support here.

Thanks - your English is great (I can't tell what parts were machine-translated).

**Comment by Jason0247** [comment]
I have seen the team's efforts, and of course, efforts should earn the benefits they deserve, so that the team can be more motivated.
In fact, for Huawei's absolute service, the team can consider using a fee like Linux to provide services. I think a better approach is to open up some features for developers to try and use the full version through payment.
I think this is a great opportunity to consider promoting Chinese developers and establishing relevant online communities. After all, Huawei's mobile phone market in China alone is already very large, so don't ignore China as a development group. After all, during the popularity of Flash, there were a large number of Chinese developers, and I think they are also willing to return.
thanks.
@ajwfrost

**Reply by shaucky** [reply]
Uh… The comment remind me something about Unity… See what Unity do for Chinese developesr: https://unity.cn/tuanjie/tuanjieyinqing .
From their introduce, Unity is ready for deployment plan on HarmonyOS (the OpenHarmony Project? )… I think the competition between different platforms on HarmonyOS will be fierce. If AIR supports HarmonyOS, it may be difficult to become an advantage. But if AIR does not support HarmonyOS, it may become a disadvantage.

**Comment by Jason0247** [comment]
According to the news, this new system may be announced in about a year, and most Chinese manufacturers are likely to no longer use the Android system.
I would like to know if AIR will have a clear plan, such as temporarily not considering support or within the support plan, which can also give us developers a clear goal of whether to switch to other platforms in advance when facing future new system challenges.
When AIR has a clear planned task, we developers have enough confidence and expectations, even if it requires a license.
Thank you. @ajwfrost

**Comment by ajwfrost** [comment]
Thanks for all the inputs!

I think we'd need to have a bit of a considered approach on this - I don't want to say now "oh yes we'll definitely support it", because we're not yet sure what it would entail. But, I will get one of the team to investigate it and to see what it would take to have support for this; if it's feasible then we can make a decision and I'll then be able to give a firm picture on the support and likely timescales..

Thanks!

   Andrew

**Comment by Jason0247** [comment]
Thank you to the AIR team for their efforts, so that AIR can receive updates and support after Adobe.
The mainstream development tools Unity and UE are indeed popular, but it cannot be denied that AIR has its own development space, and cross platform is a very important feature of it. I hope that the AIR team can seriously consider this decision, and I believe that this decision will definitely affect everyone who uses AIR.

**Comment by Jason0247** [comment]
Hello, may I ask if we have any new plans and developments for the new system?
I really need to know the future planning for the new system. Thank you @ajwfrost

**Comment by Jason0247** [comment]
It feels like everyone doesn't really care about this issue. Am I the only one paying attention.

**Reply by ajwfrost** [reply]
Hi - sorry, I've missed a lot of emails having been off to Adobe MAX and then having a few days off last week... but to update here, we are looking into the support of Harmony OS, I will need to catch up with a colleague (who's off this week) to see where we're at in terms of progress for this. I should be able to provide an update here by 1st Nov..

thanks

**Comment by Ender22** [comment]
Also awaiting an update, hoping for good news. As a b2b app with a lot of Chinese clients, this is a terrifying issue.

**Comment by Jason0247** [comment]
Hello, have you received any news lately? I have been looking forward to it. @ajwfrost

**Reply by ajwfrost** [reply]
Hi - sorry, meant to update you last week.. we've investigated, there are a few issues we're having to resolve with our IT dept to get hold of everything we need, but we are planning on going ahead with this. Not entirely sure of the timescales yet though.. the main issue we can see to resolve, as mentioned in the thread, is around the packaging and deployment.

Will try to provide updates when we have got a little further here..

**Reply by Ender22** [reply]
That's excellent news!

**Reply by Jason0247** [reply]
Hard work, come on, it's worth looking forward to. Thank you!

**Comment by Jason0247** [comment]
Hello, sorry to bother you.
May I ask if there have been any latest developments recently? @ajwfrost

**Reply by ajwfrost** [reply]
Hi - we had some initial issues getting everything set up, but it looks like we are all primed and ready to try to get this working, we even have a HarmonyOS tablet now.. my colleague did comment that there didn't seem to be a huge amount of documentation available for the sort of building/packaging that we'll need to do, but hopefully it's a similar enough platform to Android that we can work it out.

So there will be more of a focus on this when we're back in January and hopefully we'll be able to get things building and deployed, and will then see how we may need to create/adjust the ADT and packaging tools for this platform.

thanks

**Comment by Jason0247** [comment]
@ajwfrost Hello, may I ask if there is any latest progress? I'm looking forward to it.

**Reply by ajwfrost** [reply]
Hi - we're all set up with the build environments etc now .. the challenge though is that the documentation is quite sparse and no real examples of what we would need to do to get AIR running on it. This (version 3) It seems more set up as a JavaScript/webapp type mechanism - maybe with the Java interfaces it might be possible, but it sounds like a completely new platform set-up will be required, rather than being able to tweak/adapt the Android variant.

If they don't support any of the Android APIs then it's going to be quite a task to implement this. And there may be some things that are impossible - such as currently, there don't seem to be any native APIs for calling OpenGL ES functions. If we need to use JNI calls the whole time, this is going to be prohibitively slow. I'm wondering whether we need to wait for the OS to mature a little and to gain more implementations on the 'native' side in order for it to support AIR.

We'll try to progress this still, perhaps as more of a background task though, and there may well be functionality that we can't provide yet (such as Stage3D)...

thanks

**Reply by wmtj** [reply]
'we're all set up with the build environments etc now ' -- I have to say, this is really sad news, it took almost half a year ...

As far as I know, Huawei is helping a large number of well-known apps to adapt HarmonyOS on a one-to-one basis. If you really want to adapt it, you should not only read the online documents, but also contact Huawei as soon as possible, so that you can get more information and the latest HarmonyOS Next (5.0) development version, as well as more helpful assistance. 

HarmonyOS Next (5.0) is not an immature system, and hundreds of well-known apps in China have already adapted.
@ajwfrost

**Reply by Devin-Licastro** [reply]
Yeah, unfortunately, Harman putting it as a background task feels like a bit of a disaster for my company. We are b2b employee training app and have clients with properties inside of China. 100% for sure, their staff members will buy these new HarmonyOS phones, and our app won't work, and our company will take the blame and likely lose some of our clients. 
I lived in China for years and they are famous for not providing anything in English, I bet tons of Chinese-only documentation exists for this.

**Reply by ajwfrost** [reply]
Am I missing something here? "HarmonyOS Next 5.0" isn't actually out yet, according to Wikipedia?

> On August 4, 2023, the developer preview version of HarmonyOS NEXT was opened for cooperating enterprise developers, and on January 18, 2024, to the rest of developers to build and test native mobile apps. The beta version was expected to be launched in Q2 2024, and the commercial version to be made available in Q4 2024.

> The beta version of HarmonyOS Next will be launched in Q2 of 2024 to all unregistered public developers

And it still looks like this is firmly targeting web-type applications developed using TypeScript .. even the current 3.0/3.1 stuff; it has 'native' capabilities but that's still a way to process JavaScript objects.. would need a very different set-up to how even the Android functionality works with JNI:

https://developer.harmonyos.com/en/docs/documentation/doc-guides-V3/napi-guidelines-0000001493903956-V3

So - lots of applications may already be ported across to this, but I suspect they're not primarily C++ based. It sounds like this would essentially mean a completely new port of AIR, rather than being just an adaptation of the packaging and app launching system from Android to HarmonyOS.

Having said all that of course .... it does actually seem like there's OpenGL ES 3.0 support, and they do have examples that are showing Java applications calling into C++ via JNI:
https://gitee.com/harmonyos/samples/tree/master/native/NativeImage
But that's from 2 years ago, and no hints/information as to whether that mechanism will still be supported in the future. They are taking away the Android-based functionality and I am wondering if that means also the Java mechanisms.

I would suspect the better approach for us would be to focus on a version of AIR SDK that could generate webassembly-based outputs, assuming that's going to work on their web runtime.

Happy to be corrected though, if you can provide links to confirmation about the ability to deploy applications that are written in Java and C++ for the future HarmonyOS versions?

thanks

**Reply by wmtj** [reply]
Thank you for your reply!   @ajwfrost

As users of AIRSDK, we don't care how AIRSDK adapt HarmonyOS, we only care if Harman promises that AIRSDK will adapt HarmonyOS and when it will be done.

So far, hundreds of well-known apps have been adapted to HarmonyOS, and more are being added every day.   Although HarmonyOS NEXT is not yet an official release, it is enough to meet the requirements of adaptation work, otherwise how would those hundreds of apps be adapted?

Since there are hundreds of apps that are adapted already, I think AIRSDK can and should be adapted as soon as possible, rather than as "background" task.

HarmonyOS NEXT is not officially released at this time.   It is to leave time for Apps to adapt, until the commonly used apps are adapted and the ecosystem is basically mature, and then the HarmonyOS NEXT can be released at any time.   So, please don't regard the current "HarmonyOS NEXT" is not mature enough, and by "mature" everything is too late.

After the official version of HarmonyOS NEXT is released, the adapted apps can run on the new system immediately. While the apps developed by AIRSDK, no one knows when to adapt, and even do not know whether they can be adapted.

Half a year has passed, and Harman should at least be able to tell us whether it can be adapted, Yes or No?

Last year, I expected a version of AIRSDK that basically met HarmonyOS packaging requirements would be available in 2024. However, from the current situation, it is difficult to say whether we can get such an AIRSDK version by 2025 or even 2026. It is even possible to get an answer in the future that it is totally impossible at all, which would be the saddest outcome.

PS: Huawei phones will not only be sold in China...

**Reply by wmtj** [reply]
The following is a random search of the major companies to adapt to HarmonyOS news, a lot of many, only select a part .

https://m.toutiao.com/is/iFWmuqn2/
https://m.toutiao.com/is/iFWm43yf/
https://m.toutiao.com/is/iFWmVDrw/
https://m.toutiao.com/is/iFWmfx1A/
https://m.toutiao.com/is/iFWmG2yX/
https://m.toutiao.com/is/iFWmmRy3/
https://m.toutiao.com/is/iFWmsFus/
https://m.toutiao.com/is/iFWmVm3F/
https://m.toutiao.com/is/iFWmfWNh/
https://m.toutiao.com/is/iFWmx5wM/
https://m.toutiao.com/is/iFWmyoMW/
https://m.toutiao.com/is/iFWmH6e7/
https://m.toutiao.com/is/iFWmfx1A/
https://m.toutiao.com/is/iFWmHnD6/
https://m.toutiao.com/is/iFWm1BEB/
...

**Reply by ajwfrost** [reply]
The honest answer here:

> Half a year has passed, and Harman should at least be able to tell us whether it can be adapted, Yes or No?

We don't know. It may not be technically viable. With the information we have available, we don't know whether we can use the Java/JNI route, or whether we would have to adapt the AIR runtime into a massive extension of JavaScript... if the latter, then this would require us to rework a huge amount of code, and would rely on them exposing a bunch of native APIs - and we don't know if those are available or will continue to be available (again, a slight documentation issue here I think).

I have reached out to Huawei though, we can see whether they are able to provide clarity on what features/APIs will be available in this 'NEXT' version and what the future direction is... don't get me wrong, I would definitely like us to support it, I just don't currently know how practical that desire is...

thanks

**Reply by wmtj** [reply]
Thank you for your honest answer!

I have said so much above, mainly hope that you take the initiative to connect with Huawei, which is much more efficient than you read the few online documents (most of which may not be in English).

**Comment by Jason0247** [comment]
Are there any latest developments.

**Comment by Jason0247** [comment]
May I ask if there is any latest progress? @ajwfrost

**Reply by ajwfrost** [reply]
Hi - we'd not heard back from Huawei on this. Have just sent up a chaser request to try to get some info or a contact there. I will also ask one of my colleagues from our China offices to see if she can get any response..

**Reply by shaucky** [reply]
Hi, just out of curiosity, what about porting to Switch? Is the progress in this area ideal? Thanks.

**Reply by ajwfrost** [reply]
I need to chase them again .. held up with legal issues where they want to have some rights to the AIR SDK which we aren't allowed to give, based on the rights we hold from Adobe. Stuck in the middle a little here! But we may have a way forwards using a different mechanism.

It's all a little slow though :-(

**Reply by ajwfrost** [reply]
@shaucky we got a bit further so it does actually look like we have a route forwards now for Switch support. Next steps are to confirm efforts and sort out how this could work commercially...

thanks

**Reply by mrfrasier** [reply]
I don't even make games but this sounds awesome!

**Comment by Jason0247** [comment]
I have been closely following this progress. Is there any latest news? Thank you.@ajwfrost

**Reply by ajwfrost** [reply]
HI

We've not had any responses from Huawei to our queries, even when sent in Chinese from a colleague from our Chengdu office. So at the moment we're still unclear as to whether this is going to be possible.

Our ideal would be if we can get some sample code (and the assurances that this mechanism will remain a possibility in forthcoming versions of HarmonyOS) for a Java-based application that loads in some C++ native code via JNI. This was available in the initial versions of the OS but if the next versions will just support web-style apps that require JavaScript (or WebAssembly) then it will be more of a challenge i.e. impossible until we get a WebAssembly variant created..

thanks

**Comment by Jason0247** [comment]
@ajwfrost 
Do I understand that the team has given up on this project task?
thanks

**Reply by ajwfrost** [reply]
Not completely .. but we may need to work through another of Adobe's partners in the China region in order to progress things further here. So nothing in the short term I'm afraid.

**Comment by shaucky** [comment]
Hi, 
 
Hoped to inquire about some news.
 
Huawei is not very proactive in responding to cross platform technology.  I heard in the Avalonia (a .NET-based cross-platform framework) community a few days ago that they did not receive any response to their inquiry. So whether AIR or other cross-platform technologies, may be difficult to proactively support HarmonyOS.
 
Is there any news with AIR for new OS support, any plan for version 52.x or new language support .e.g? Also a few days ago, i saw news in the AIR community about there's no Animate 2025. Is it hopeful Harman take over Animate also? (Well… this is a self-evident question, but currently AIR if be used as a game development platform, it still lacks competitiveness in terms of tools, i think)
 
Since above. As you guys see, many people use AIR because of Animate. Although Animate outputs classic Flash assets, but its resource integration with Starling, Away3D, and others that utilize the Stage 3D framework is not good, this left people with the impression of 'the different name, the same shit'…
 
And then on the other hand, when I tried to implement RFC 4954 (if I remember correctly) using AS3 , I found that AIR's API was not sufficiently low-level. The standard requires establishing a Socket connection first and upgrading to a TLS Socket connection after certain interactions, but APIs of AIR do not support such operations.
 
Since above… If trying redesign the API, changing the language version is also necessary. I remember a long time ago, we talked about the possibility of a new version of ActionScript.  Andrew has said that Adobe has a standard for ActionScript 4.0, so Harman cannot change it and has no plans to implement it in the short term. A few months later, i suggested that since there are updates to language elements, is it possible to change the sub version number, such as 3.1. But this matter was also left unresolved. What if jumping to version 5.0? And some languages like C# use the same CLR (equivalent to AVM2) for different version instead of refactoring.
 
Have to admit that what comes to mind in the latter part is what is written. I actually understand that there are issues related to manpower and cost allocation here. Thank you for reading this. My original intention was still to ask if there is any relevant news.
 
Thanks. @ajwfrost

**Reply by ajwfrost** [reply]
Hi - just as an update around the Huawei/Harmony part of it - please check the other thread on this, #3327 which we can take forward as a more specific conversation around Harmony OS support, this thread is a bit more generic now (and I'll hopefully address some of your other queries in the not-too-distant future..)

thanks

**Reply by shaucky** [reply]
Thanks for your reading & reply. I'll keep focusing on news like HarmonyOS support and so on.
 
My intention was not to cause bother or pressure. If my emotions have caused you any trouble, i didn't mean to.
 
We all hope that AIR can become a popular technology.
